BE-006 (Nikolay Chornyi) Add entity, repository & services

mblack666 opened the pull request:
Add entity, repository & services: Category, Discount, Vendor, VendorLocation

mblack666 committed:
BE-006 (Nikolay Chornyi) Add entity, repository & services: Category, Discount, Vendor, VendorLocation

el-dub commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,40 @@
+
+    private String name;
+
+    @OneToMany(cascade = CascadeType.ALL,
+    fetch = FetchType.EAGER,
+    mappedBy = "category")
+    private List<Discount> discount;
```
maybe discounts in plural

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,40 @@
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private UUID id;
```
You can use this
@GeneratedValue(generator = "UUID")
@GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
for UUID

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -0,0 +1,59 @@
+    @JoinColumn(name = "category_id")
+    private Category category;
+
+    private String name;
+    private String description;
+    private String promo;
+    private int percent;
+    private Date start_time;
+    private Date end_time;
+    private boolean active;
```
Maybe @Column(name = "...") annotation will be good for each field, but your variables should be startTime instead_of start_time

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Vendor.java:
```diff
@@ -0,0 +1,44 @@
+    private String name;
+    private String description;
+
+    @OneToMany(cascade = CascadeType.ALL,
+    fetch = FetchType.EAGER,
+    mappedBy = "vendor")
+    private List<VendorLocation> locationVendors = new ArrayList<>();
+
```
vendorLocations instead of locationVendors

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/repository/VendoLocationrRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface VendoLocationrRepository extends JpaRepository<VendorLocation, Long> {
+}
```
Mistake in naming: VendorLocationRepository
And also we should use extends JpaRepository<VendorLocation, UUID> not Long, in each repository

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,40 @@
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return "Category{" +
+                "id=" + id +
+                ", name='" + name + '\'' +
+                '}';
+    }
```
I don't understand the aim of overriding toString() by hand. We can use Lombok's @ToString(exclude ={"someField", ...}). And the same for hachCode. You could write @Data and then @ToString(exclude = {"discounts"}), @EqualsAndHashCode (exclude = {"discounts"). It will brief than @Getter @Setter.

mblack666 replied:
+

mblack666 committed:
Use @Data Lombok, change PK Long to UUID & type generate UUID

SergeyKoval commented:
Please add @Column annotations on each column (and limitations like max length or not nullable base on db schema)

SergeyKoval reviewed (changes requested):
Please add answer to each comment you have processed. At least "done" in order to see, that you don't miss the item

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,34 @@
+
+    private String name;
+
+    @OneToMany(cascade = CascadeType.ALL,
```
Don't put cascade if you don't really need it. If you need some cascade type, put just required

mblack666 replied:
done

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,34 @@
+    private String name;
+
+    @OneToMany(cascade = CascadeType.ALL,
+    fetch = FetchType.EAGER,
```
Try to use as much lazy as possible. Don't use eager without real reason

mblack666 replied:
done

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,34 @@
+    @OneToMany(cascade = CascadeType.ALL,
+    fetch = FetchType.EAGER,
+    mappedBy = "category")
+    @ToString.Exclude
```
Why are you excluding discounts from toString?

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,34 @@
+    @ToString.Exclude
+    private List<Discount> discounts;
+
+    public Category(String name) {
```
Do you really need this constructor? For what reason?

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -0,0 +1,42 @@
+    private String description;
+    private String promo;
+    private int percent;
+    private Date start_time;
```
Don't use Date type

mblack666 replied:
+

SergeyKoval commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -0,0 +1,42 @@
+    private Date end_time;
+    private boolean active;
+
+    public Discount(Category category, String name, String description, String promo, int percent, Date start_time, Date end_time, boolean active) {
```
Remove all such constructors from entities

mblack666 replied:
done

n-chornyi replied:
+

SergeyKoval commented on src/main/java/com/exadel/discount/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+@Service
+public class CategoryServiceImpl implements CategoryService {
+
+    @Autowired
```
Use constructor injection instead

mblack666 replied:
+

mblack666 committed:
Change field constructor to injection constructor
Remove all such constructors
Срфтпу fetch.EAGER to fetch.LAZY
Change Cascade
Add @Column

el-dub commented on src/main/java/com/exadel/discount/repository/VendoLocationrRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.UUID;
+
+@Repository
+public interface VendoLocationrRepository extends JpaRepository<VendorLocation, UUID> {
```
Mistake in name

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/VendorLocation.java:
```diff
@@ -0,0 +1,42 @@
+    private Vendor vendor;
+
+    @Column(name = "country_id")
+    private UUID country_id;
+    @Column(name = "city_id")
+    private UUID city_id;
```
please, rename this columns in camel case countryId, cityId

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -0,0 +1,44 @@
+    private String promo;
+    @Column(name = "percent")
+    private int percent;
+    @Column(name = "start_time", nullable = false)
+    private ZonedDateTime start_time;
```
start_time can be nullable according to outr DB

mblack666 replied:
+

Hanna-Kalashnyk replied:
The time in Coupone entity is in LocalDateTime type. Should they to be the same?

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Column(name = "name", length = 50, nullable = false)
+    private String name;
+    @Column(name = "description", length = 255, nullable = false)
+    private String description;
```
Can be null according to our DB

mblack666 replied:
+

mblack666 committed:
edit column entity, rename VendoLocationrRepository

mblack666 committed:
endTime may be nullable

el-dub committed:
DB changes

el-dub committed:
DB changes fixes

el-dub committed:
Merge branch 'dev' into entityv2

mblack666 committed:
.

SergeyKoval commented on src/main/java/com/exadel/discount/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public Category get(UUID id) {
+        return categoryRepository.getOne(id);
```
Use findById instead. Difference you can read here: https://www.javacodemonk.com/difference-between-getone-and-findbyid-in-spring-data-jpa-3a96c3ff

mblack666 replied:
+

SergeyKoval commented on src/main/resources/db/changelog/v.0.2/db.changelog-drop-not-null-constraint.yaml:
```diff
@@ -0,0 +1,8 @@
+databaseChangeLog:
+  - changeSet:
+      id: 29-drop-nullable-true-in-discounts-end-time
+      author: elizaveta_dubogryz
```
looks like you are not Elizaveta

SergeyKoval commented on src/main/java/com/exadel/discount/service/VendorServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public Vendor get(UUID id) {
+        return vendorRepository.getOne(id);
```
Convert to DTO on service layer. Ask @el-dub to provide knowledge transfer for the DTO transformation, looks like she has already covered this topic.

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/entity/VendorLocation.java:
```diff
@@ -0,0 +1,42 @@
+    private String contact;
+    @Column(name = "coordinates")
+    private String coordinates;
+
```
Exactly the same column name postgresql makes itself automatically.
Think , annotations @Column( are redundant in some  fields.

mblack666 replied:
I think either all annotations or none

Hanna-Kalashnyk replied:
Agree) Done

Hanna-Kalashnyk replied:
U're right. So it's decided to annotate every field.

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/service/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public Discount get(UUID id) {
+        return discountRepository.getOne(id);
+    }
```
Would U change getOne(id) to FindById(id)?

mblack666 replied:
+

mblack666 committed:
Add mapper, dto, edit service & entity

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/dto/DiscountDTO.java:
```diff
@@ -0,0 +1,39 @@
+    @NotBlank(groups = Create.class)
+    private String name;
+    @Null(groups = Create.class)
+    private String description;
```
Does it mean, we could not add description at discountDTO creation? Maybe, no annotation?

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/dto/DiscountDTO.java:
```diff
@@ -0,0 +1,39 @@
+    @NotNull(groups = Create.class)
+    private int percent;
+    @Null(groups = Create.class)
+    private LocalDateTime startTime;
```
may be , Keep the ability to put start time at. creation: no annotation at field?

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/dto/DiscountDTO.java:
```diff
@@ -0,0 +1,39 @@
+    @NotNull(groups = Create.class)
+    private LocalDateTime endTime;
+    @NotNull(groups = Create.class)
+    private boolean active;
```
endTime could be not set at creation too ,as I've understood at meeting from FE... Ask Lisa.

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,37 @@
+    private UUID id;
+    @Column(name = "name", nullable = false, length = 50)
+    private String name;
+
+    @OneToMany(cascade =  CascadeType.REMOVE,
+            mappedBy = "category")
+    private List<Discount> discounts;
```
I don't think that it should be REMOVE

mblack666 replied:
What category will be in discount after delete category? (Category cant't be null)

Hanna-Kalashnyk replied:
Nikolay meant, all discounts of the removed category will be deleted also together with category deleting. It's a question to PO.

el-dub commented on src/main/java/com/exadel/discount/dto/VendorLocationDTO.java:
```diff
@@ -0,0 +1,26 @@
+public class VendorLocationDTO {
+    @Null(groups = Create.class)
+    private UUID id;
+    @NotNull(groups = Create.class)
+    private Vendor vendor;
```
VendorDTO maybe

mblack666 replied:
-
deleted field

el-dub commented on src/main/java/com/exadel/discount/dto/VendorDTO.java:
```diff
@@ -0,0 +1,30 @@
+    @Null(groups = Create.class)
+    private String description;
+    @Null(groups = Create.class)
+    private List<VendorLocation> vendorLocations;
```
I agree with Anna. Think about it, I think that it could be not null, when we create new Vendor

mblack666 replied:
i think it's null, first create vendor, after add location

Hanna-Kalashnyk replied:
FE really showed us two-stage mechanism of vendor creation
and I agree with
 @Null(groups = Create.class)
    private List<VendorLocation> vendorLocations;

But field in Vendor.class
 @Column(name = "description")
    private String description;

Could be not null at creation. 
Maybe, Without @Null(groups = Create.class) befor description?

mblack666 replied:
description may be null at creation

mblack666 committed:
Add controller Vendor, VendorLocation, Category, Discount
Edited service, dto

Hanna-Kalashnyk committed:
Merge branch 'dev' into entityv2

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public CategoryDTO get(UUID id) {
+        return categoryMapper.getDTO(categoryRepository.findById(id).orElse(null));
```
Project style is with explicit NotFoundException masseging, as well as loggingn with sjt4j.
Like here:
@Override
    public CategoryDTO get(UUID id) {
        log.debug("Finding Category by ID");

        Category category = categoryRepository
                .findById(id)
                .orElseThrow(() -> new NotFoundException(String.format("Category with id %s not found", id)));

        log.debug("Successfully Category is found by ID");
        return categoryMapper.toCategorynDto(category);
    }

mblack666 replied:
+

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/service/VendorLocationServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+                .getVendor()
+                .getId()
+                .equals(vendorId)) {
+            vendorLocationRepository.deleteById(locationId);
```
Are you sure about "throws Exception" up?
May be Should here to put any massege , if  notequals(vendorId)?
Like this:
else throw new NotFoundException(.....     or something...
Elso logging.

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -16,37 +16,64 @@
-@EqualsAndHashCode(exclude = {"tags"})
-@ToString(exclude = {"tags"})
+@EqualsAndHashCode(exclude = {"tags", "coupons", "favorites"})
+@ToString(exclude = {"tags", "coupons", "favorites"})
```
Hope, U do not mind my edition... )

el-dub replied:
I think adding something that doesn't work to dev is not a good idea. Nikolay's PR doesn't have Coupon and Favorite. It will better to add these fields later with entities that respond for them

el-dub replied:
And I also think it's not good to perform merging dev into not yours branch. As a result we have not working code

mblack666 replied:
I agree with Elizaveta, add part code with class, which i don't have, it's very bad idea.
Also about merge dev to my branch, don't make this more.

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/service/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    public List<DiscountDTO> getAll() {
+        return discountMapper.getListDTO(discountRepository.findAll());
+    }
```
Would U add any logs?

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -1,6 +1,6 @@
 package com.exadel.discount.entity;
 
-import lombok.AllArgsConstructor;
+mport lombok.AllArgsConstructor;
```
Mistake

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -16,37 +16,64 @@
+  
+    @Column(name = "active")
+    private boolean active;
+  
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "category_id")
```
ManyToOne import is missing

mblack666 replied:
This is because of the merge, i have it

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/service/VendorServiceImpl.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public void delete(UUID id) {
+        vendorRepository.deleteById(id);
```
Do we need here NotFound Exception?
Also 
@Transactional
for thread-saving work. As at every DB changing methods.

Hanna-Kalashnyk commented:
I hope to merge today these fields. So I have added them. Let ask mentors
about sequence of feature adding.
These are my changes in his Discount.java. Sorry.

чт, 24 июн. 2021 г. в 09:05, Elizaveta Dubogryz ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/exadel/discount/entity/Discount.java
> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r657650806>
> :
>
> >  @NoArgsConstructor
>  @AllArgsConstructor
> -@EqualsAndHashCode(exclude = {"tags"})
> -@ToString(exclude = {"tags"})
> +@EqualsAndHashCode(exclude = {"tags", "coupons", "favorites"})
> +@ToString(exclude = {"tags", "coupons", "favorites"})
>
> I think adding something that doesn't work to dev is not a good idea.
> Nikolay's PR doesn't have Coupon and Favorite. It will better to add these
> fields later with entities that respond for them
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r657650806>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMP2Q7IJQJIPQU4OD3P5O3TULDLJANCNFSM46QKRRJQ>
> .
>

el-dub commented on src/main/java/com/exadel/discount/controller/DiscountController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping
+    @ApiOperation("Add new discount")
+    public DiscountDTO addDiscount(@RequestBody @Valid DiscountDTO discountDTO) {
+        return discountService.save(discountDTO);
+    }
+
```
Validation doesn't work. Change @Valid to @Validated(Create.class)

mblack666 replied:
+

Hanna-Kalashnyk commented:
ok. Sorry, Nikolay.

чт, 24 июн. 2021 г. в 09:38, Hanna Kalashnyk ***@***.***>:

> I hope to merge today these fields. So I have added them. Let ask mentors
> about sequence of feature adding.
> These are my changes in his Discount.java. Sorry.
>
> чт, 24 июн. 2021 г. в 09:05, Elizaveta Dubogryz ***@***.***
> >:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In src/main/java/com/exadel/discount/entity/Discount.java
>> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r657650806>
>> :
>>
>> >  @NoArgsConstructor
>>  @AllArgsConstructor
>> -@EqualsAndHashCode(exclude = {"tags"})
>> -@ToString(exclude = {"tags"})
>> +@EqualsAndHashCode(exclude = {"tags", "coupons", "favorites"})
>> +@ToString(exclude = {"tags", "coupons", "favorites"})
>>
>> I think adding something that doesn't work to dev is not a good idea.
>> Nikolay's PR doesn't have Coupon and Favorite. It will better to add these
>> fields later with entities that respond for them
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r657650806>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AJMP2Q7IJQJIPQU4OD3P5O3TULDLJANCNFSM46QKRRJQ>
>> .
>>
>

el-dub commented on src/main/java/com/exadel/discount/dto/DiscountDTO.java:
```diff
@@ -0,0 +1,31 @@
+
+@Data
+public class DiscountDTO {
+    @Null(groups = Create.class)
+    private UUID id;
+    @NotNull(groups = Create.class)
+    private CategoryDTO category;
+    @NotBlank(groups = Create.class)
+    private String name;
```
Add messages for validation errors please

mblack666 replied:
+

mblack666 committed:
Add logger, minor edits

mblack666 committed:
Merge remote-tracking branch 'origin/entityv2' into entityv2

# Conflicts:
#	src/main/java/com/exadel/discount/entity/Discount.java

mblack666 committed:
Add ManyToMany Tags in Discount

el-dub commented on src/main/java/com/exadel/discount/dto/DiscountDTO.java:
```diff
@@ -0,0 +1,31 @@
+    @NotBlank(groups = Create.class, message = "Discount promo should be not blank")
+    @Size(max = 50, groups = Create.class, message = "Discount promo should be shorted than {max}")
+    private String promo;
+    @NotNull(groups = Create.class)
+    private int percent;
```
NotNull won't work with int. And according to our DB it can be null

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -1,46 +1,61 @@
 @Data
 @NoArgsConstructor
-@AllArgsConstructor
-@EqualsAndHashCode(exclude = {"tags"})
-@ToString(exclude = {"tags"})
```
Why did you delete this?

Hanna-Kalashnyk replied:
maybe, to add this
@EqualsAndHashCode(exclude = {"tags", "coupons", "favorites", "vendorLocations"})
@ToString(exclude = {"tags", "coupons", "favorites", "vendorLocations"})

;)

mblack666 replied:
Added

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/DiscountController.java:
```diff
@@ -0,0 +1,45 @@
+
+@Validated
+@RestController
+@RequiredArgsConstructor
```
Ad d this, please
@EqualsAndHashCode(exclude = {"tags", "coupons", "favorites"})
@ToString(exclude = {"tags", "coupons", "favorites"})

mblack666 replied:
No, will be add after yout PR

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/DiscountController.java:
```diff
@@ -0,0 +1,45 @@
+    public DiscountDTO addDiscount(
+            @RequestBody @Validated(Create.class) DiscountDTO discountDTO) {
+        return discountService.save(discountDTO);
+    }
```
Remove spare lines, please

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/VendorController.java:
```diff
@@ -1,90 +1,76 @@
 
 @Validated
 @RestController
+@RequiredArgsConstructor
```
@EqualsAndHashCode(exclude
@ToString(exclude
are needed too
for running

mblack666 replied:
Why?

Hanna-Kalashnyk replied:
It does not work without them at my last Idea. It's counting hashCodes all the time with real entity values.

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/VendorController.java:
```diff
@@ -1,90 +1,76 @@
+        return vendorService.getById(id).getVendorLocations();
     }
 
     @PostMapping("/{vendorId}/locations")
```
no PathVariables at Put and Post requests mentors demand according to REST article.
Change to @RequestParam
Or could write CreateVendorLocationDto, witch will contain both arguments and pass it in @RequestBody

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -1,46 +1,61 @@
-import lombok.EqualsAndHashCode;
-import lombok.NoArgsConstructor;
-import lombok.ToString;
+import lombok.*;
```
No wildCards allowed

mblack666 replied:
+

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/entity/Category.java:
```diff
@@ -0,0 +1,37 @@
+package com.exadel.discount.entity;
+
+import lombok.*;
```
mentors will ask to remove wilds.. sure

mblack666 replied:
+

Hanna-Kalashnyk commented:
Could U rewatch last time and throw into the mentor chat for approval,
please. The deadline is close.

чт, 24 июн. 2021 г. в 18:50, n-chornyi ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/exadel/discount/controller/VendorController.java
> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r658073130>
> :
>
> >
>  @Validated
>  @RestController
> +@RequiredArgsConstructor
>
> Why?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r658073130>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMP2Q4BYZIPECBDUOIGH5LTUNH3NANCNFSM46QKRRJQ>
> .
>

Hanna-Kalashnyk commented:
I know U're busy, but , please:
Rewatch all comments and fix all which are not yet, please. It's needed
now...

чт, 24 июн. 2021 г. в 19:26, Hanna Kalashnyk ***@***.***>:

> Could U rewatch last time and throw into the mentor chat for approval,
> please. The deadline is close.
>
> чт, 24 июн. 2021 г. в 18:50, n-chornyi ***@***.***>:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In src/main/java/com/exadel/discount/controller/VendorController.java
>> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r658073130>
>> :
>>
>> >
>>  @Validated
>>  @RestController
>> +@RequiredArgsConstructor
>>
>> Why?
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/SergeyKoval/2021_june_team4_be/pull/12#discussion_r658073130>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AJMP2Q4BYZIPECBDUOIGH5LTUNH3NANCNFSM46QKRRJQ>
>> .
>>
>

mblack666 committed:
Change PathVariables to RequestParam
Remove wildCards
Made out a separate class VendorLocationController

mblack666 committed:
Clear code

el-dub commented on src/main/java/com/exadel/discount/service/interfaces/VendorService.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface VendorService {
+
+    VendorDTO save(VendorDTO vendor);
+    VendorDTO getById(UUID id);
+    List<VendorDTO> getAll();
+    void deleteById(UUID id);
```
Format your code with ctrl + alt +L

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/service/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+package com.exadel.discount.service;
+
```
I think we should keep single project structure. For example, all impl in impl package. But now we have some in impl, some out of it. The same is about interfaces package.

mblack666 replied:
Yes, I already changed structure in my project but don't push yet.

el-dub commented on src/main/java/com/exadel/discount/entity/Discount.java:
```diff
@@ -1,46 +1,66 @@
+
+    /*@ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "discount_locations",
```
Our table has name discounts_locations

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/controller/CategoryController.java:
```diff
@@ -0,0 +1,55 @@
+    private final CategoryService categoryService;
+
+    @GetMapping
+    @ApiOperation("Get all categories or category by id")
+    public List<CategoryDTO> getAll(@RequestParam(name = "id", required = false) UUID id) {
+        if (id == null) {
```
I don't think that combining getAll and getById is a good idea. The name of your method is getAll, it's not good to find by id in it. And also for other entities the same. One more reason to devide these methods is, for example, for discounts: in getAll we'll pass some sorting, paging parameters

mblack666 replied:
+

mblack666 committed:
PathVariable, edited service, controller, entity

mblack666 committed:
Add auto-generate ID in table discounts_locations

mblack666 committed:
comment don't merge code, move auto-generate id to v.0.3

mblack666 committed:
{Delete} Add error when don't find (category, vendor, vendorlocation, cendor) by id

mblack666 committed:
minor edits

mblack666 committed:
minor edits

SergeyKoval commented on src/main/java/com/exadel/discount/controller/VendorController.java:
```diff
@@ -1,90 +1,91 @@
+    @GetMapping("/{vendorId}/locations")
+    @ApiOperation("Get list of all vendor locations")
+    public List<VendorLocationDTO> getVendorLocations(@PathVariable(name = "vendorId") @NotNull UUID id) {
+        return vendorService.getById(id).getVendorLocations();
```
Why don't you get vendor locations directly from repository?

mblack666 replied:
+

SergeyKoval commented on src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+        }
+        discount.setVendorLocations(vendorLocations);*/
+        DiscountDTO discountDTO = discountMapper.getDTO(discount);
+        if (discount.getVendorLocations().get(0) != null) {
```
What is going on here??? Looks vary strange

mblack666 replied:
+

SergeyKoval commented on src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+        log.debug("Getting list of all Discounts");
+        List<Discount> discounts = discountRepository.findAll();
+        //TODO
+        /*for (int i=0; i<discounts.size(); i++) {
```
This should be populated on db call or I don't understand the logic...

mblack666 replied:
+

SergeyKoval commented on src/main/java/com/exadel/discount/service/impl/VendorLocationServiceImpl.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Override
+    public void deleteById(UUID vendorId, UUID locationId)  {
```
Logic of this method should be refactored

mblack666 replied:
+

mblack666 committed:
Delete Service, Update controllers, move VendorLocationController to new class

mblack666 committed:
returned service

mblack666 committed:
.

el-dub commented on src/main/java/com/exadel/discount/controller/CategoryController.java:
```diff
@@ -0,0 +1,61 @@
+        return categoryService.save(categoryDTO);
+    }
+
+    @DeleteMapping
+    @ApiOperation("Delete category by Id")
+    public void deleteCategory(@RequestParam(name = "id", required = true) @NotNull UUID id) {
+        categoryService.deleteById(id);
+    }
```
I think here and in other delete methods id should be a PathVariable

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/controller/VendorLocationController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @PostMapping
+    @ApiOperation("Add new location of vendor")
+    public VendorLocationDTO addVendorLocation(@RequestParam(name = "vendorId", required = true) @NotNull UUID vendorId,
+                                               @RequestBody @Validated VendorLocation vendorLocation) {
+        return vendorLocationService.save(vendorLocation, vendorId);
+    }
```
RequestBidy should be VendorLocationDto and I think that we don't need RequestParam, vendorId should be in vendorLocationDto

mblack666 replied:
+

mblack666 committed:
Changed RequestParam to PathVariable in DeleteMapping

el-dub commented on src/main/java/com/exadel/discount/service/impl/VendorLocationServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public VendorLocationDTO save(VendorLocation vendorLocation, UUID vendorId) {
+        log.debug("Saving new VendorLocation");
+        Vendor vendor = new Vendor();
+        vendor.setId(vendorId);
```
I think here we should get Vendor by id from vendorRepository and then set it to vendorLocation

mblack666 replied:
Its work

el-dub commented on src/main/java/com/exadel/discount/service/impl/VendorLocationServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public List<VendorLocationDTO> getAll(UUID id) {
+        log.debug("Getting list of all Vendors");
+        Vendor vendor = new Vendor();
+        vendor.setId(id);
```
The same about vendor is here. And if we haven't found this vendor by id, it should throw an exception

mblack666 replied:
+

mblack666 committed:
Add dto for create vendor & location

el-dub committed:
Update DiscountRepository.java

el-dub committed:
VendorLocation updates

el-dub committed:
Update validation

el-dub committed:
Add many changes to dicsount, vendor, location services

el-dub committed:
Update Discount.java

SergeyKoval commented on src/main/java/com/exadel/discount/mapper/DiscountMapper.java:
```diff
@@ -0,0 +1,37 @@
+
+    Discount parseDTO(CreateDiscountDTO discounts);
+
+    default BaseVendorDTO getVendor(Discount discount) {
```
Very strange method!!! it should not be like this!

mblack666 replied:
+

mblack666 committed:
Add communication with discount & vendor (ManyToOne), add create field and FK to DB in LiquiBase

mblack666 committed:
Edits discountDTO & service

el-dub commented on src/main/java/com/exadel/discount/mapper/DiscountMapper.java:
```diff
@@ -0,0 +1,23 @@
+
+    Discount parseDTO(CreateDiscountDTO discounts);
+
+    BaseVendorDTO getDTO(Vendor vendor);
```
Delete this

mblack666 replied:
+

el-dub commented on src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private Set<VendorLocation> findVendorLocations(UUID vendorId, List<UUID> locationIds) {
+        /*vendorRepository.findById(vendorId)
+                .orElseThrow(() -> new NotFoundException(String.format("Vendor with ID %s doesn't exist", vendorId)));*/
```
Remove this

mblack666 replied:
+

mblack666 committed:
Deleted unused methods

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/CategoryController.java:
```diff
@@ -0,0 +1,56 @@
+    @ApiOperation("Get information about category by Id")
+    public CategoryDTO getById(
+            @PathVariable(name = "categoryId") @NotNull UUID id) {
+        return categoryService.getById(id);
```
.orElseThrow(new NotFoundException(String.format("Category with id %s not found", id));
maybe here, or something...

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/CategoryController.java:
```diff
@@ -0,0 +1,56 @@
+    @ApiOperation("Delete category by Id")
+    public void deleteCategory(@PathVariable(name = "id") @NotNull UUID id) {
+        categoryService.deleteById(id);
+    }
```
NotFoundException will be good here

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/DiscountController.java:
```diff
@@ -0,0 +1,57 @@
+    @ApiOperation("Get discount by ID")
+    public DiscountDTO getDiscountById(
+            @PathVariable(name = "discountId") @NotNull UUID id) {
+        return discountService.getById(id);
```
NotFoundException will be good here

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/DiscountController.java:
```diff
@@ -0,0 +1,57 @@
+    @DeleteMapping("/{id}")
+    @ApiOperation("Delete discount")
+    public void deleteDiscount(@PathVariable(name = "id") @NotNull UUID id) {
+        discountService.deleteById(id);
```
NotFoundException with special messege will be good here
Or default can leave...

Hanna-Kalashnyk commented on src/main/java/com/exadel/discount/controller/VendorLocationController.java:
```diff
@@ -0,0 +1,57 @@
+    @GetMapping("/{id}")
+    @ApiOperation("Get location by ID")
+    LocationDTO getById(@PathVariable(name = "id") @NotNull UUID id) {
+        return vendorLocationService.getById(id);
```
NotFoundException will be good here
